internal/login/login.go

```diff
@@ -100,19 +107,62 @@ func FetchDeviceAuthorization(
 	return deviceAuthorization, nil
 }
 
+// FetchToken attempts to get an access token. It will continue to try while the user logs in to
+// verify their request. If the user denies the request or does nothing long enough for this call
+// to time out, we do not return an access token.
 func FetchToken(
 	client UnauthenticatedClient,
 	deviceCode string,
 	baseURI string,
+	interval time.Duration,
+	maxAttempts int,
+) (DeviceAuthorizationToken, error) {
+	var attempts int
+	for {
+		if attempts >= maxAttempts {
+			return DeviceAuthorizationToken{}, errors.NewError("The request timed-out after too many attempts.")
+		}
+		deviceAuthorizationToken, err := fetchToken(
+			client,
+			deviceCode,
+			baseURI,
+		)
+		if err == nil {
+			return deviceAuthorizationToken, nil
+		}
+
+		var e struct {
+			Code    string `json:"code"`
+			Message string `json:"message"`
+		}
+		err = json.Unmarshal([]byte(err.Error()), &e)
+		if err != nil {
+			return DeviceAuthorizationToken{}, errors.NewErrorWrapped("error reading response", err)
+		}
+		switch e.Code {
+		case "authorization_pending":
+			attempts += 1
+		case "access_denied":
+			return DeviceAuthorizationToken{}, errors.NewError("Your request has been denied. Please try logging in again.")
+		case "expired_token":
+			return DeviceAuthorizationToken{}, errors.NewError("Your request has expired. Please try logging in again.")
+		default:
+			return DeviceAuthorizationToken{}, errors.NewErrorWrapped("We cannot complete your request.", err)
+		}
+		time.Sleep(interval)
+	}
+}
+
+func fetchToken(
+	client UnauthenticatedClient,
+	deviceCode string,
+	baseURI string,
 ) (DeviceAuthorizationToken, error) {
 	path := fmt.Sprintf("%s/internal/device-authorization/token", baseURI)
-	body := fmt.Sprintf(
-		`{
-			"deviceCode": %q
-		}`,
-		deviceCode,
-	)
-	res, err := client.MakeRequest("POST", path, []byte(body))
+	body, _ := json.Marshal(map[string]string{
```

Comment:
Marshal this instead of using a string since the test mock complains if the whitespace differs from the test setup. Marshaling removes extra whitespace to match the tests